### ISSUE TITLE
Restructure some concepts in ElasticDL

### ISSUE THREAD

**Issue by wangkuiyi** [sql-machine-learning/elasticdl#275]
I used to think that deep learning programmers would like to provide a function `forward` just like users of Hadoop MapReduce who like to provide `map` and `reduce`. However, deep learners would need to provide `loss`, `model`, and `optimizer` to train a model.

Obviously, `loss` calls `forward`; but who is gonna provide `forward`? I think the modeling team, who provide a set of `.py` files of Keras classes derived from `tf.keras.Model`.

To illustrate my idea, I attach a runnable TensorFlow 2.0 alpha (eager mode) program for your reference:

```python
#!/usr/bin/env python

import tensorflow as tf


# Supposed to be in the model base: github.com/sqlflow/models.
class Model(tf.keras.Model):
    def __init__(self):
        super(Model, self).__init__()
        self.W = tf.Variable(5., name='weight')
        self.B = tf.Variable(10., name='bias')

    def call(self, inputs):
        return inputs * self.W + self.B


# The loss function, model, and optimizer are from users of ElasticDL.
def loss(model, record_str):
    ns = [float(s) for s in record_str.split()]
    error = model(ns[0]) - ns[1]
    return tf.reduce_mean(tf.square(error))


model = Model()  # or, model = sqlflow.models.Model()
optimizer = tf.optimizers.SGD(learning_rate=0.01)


def grad(model, record_str):  # The grad function is part of ElasticDL.
    with tf.GradientTape() as tape:
        loss_value = loss(model, record_str)
    return tape.gradient(loss_value, [model.W, model.B])


for i in xrange(300):  # The training loop is part of ElasticDL
    with open("./data.txt") as fp:
        for cnt, line in enumerate(fp):
            grads = grad(model, line)
            optimizer.apply_gradients(zip(grads, [model.W, model.B]))
        if i % 100 == 0:
            print("Loss at epoch {:03d}: {:.3f}".format(
                i, loss(model, line)))


print("W = {}, B = {}".format(model.W.numpy(), model.B.numpy()))
```

To run the above program, say, named `a.py`, please run the following command line:

```bash
docker run --rm -it -v $PWD:/work -w /work tensorflow/tensorflow:2.0.0a0 python a.py
```

**Comment by ywskycn** [comment]
This should be already covered by existing model_zoo and model specification.
